pkg/resource/ignored_test.go

```diff
@@ -0,0 +1,271 @@
+/*
+Copyright 2023 Upbound Inc.
+*/
+
+package resource
+
+import (
+	_ "embed"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestGetIgnoredFields(t *testing.T) {
+	type args struct {
+		initProvider map[string]any
+		forProvider  map[string]any
+	}
+	type want struct {
+		ignored []string
+	}
+	cases := map[string]struct {
+		reason string
+		args
+		want
+	}{
+		"InitProviderEmpty": {
+			reason: "Should return the empty ignored fields when initProvider is empty",
+			args: args{
+				forProvider: map[string]any{
+					"singleField": "bob",
+				},
+			},
+			want: want{
+				ignored: []string{},
+			},
+		},
+		"ForProviderEmpty": {
+			reason: "Should return the ignored fields when forProvider is empty",
+			args: args{
+				initProvider: map[string]any{
+					"singleField": "bob",
+				},
+			},
+			want: want{
+				ignored: []string{"singleField"},
+			},
+		},
+		"SingleField": {
+			reason: "Should return the single ignored field when forProvider has a different field",
+			args: args{
+				initProvider: map[string]any{
+					"singleField": "bob",
+				},
+				forProvider: map[string]any{
+					"otherField": "bob",
+				},
+			},
+			want: want{
+				ignored: []string{"singleField"},
+			},
+		},
+		"SingleFieldSame": {
+			reason: "Should not return the single ignored field when forProvider has same fields",
+			args: args{
+				initProvider: map[string]any{
+					"singleField": "bob",
+				},
+				forProvider: map[string]any{
+					"singleField": "bob",
+				},
+			},
+			want: want{
+				ignored: []string{},
+			},
+		},
+		"NestedArrayField": {
+			reason: "Should return the ignored field when init and for provider have nested arrays with different fields",
+			args: args{
+				initProvider: map[string]any{
+					"array": []any{
+						map[string]any{
+							"singleField": "bob",
+							"same":        "same",
+						},
+					},
+				},
+				forProvider: map[string]any{
+					"array": []any{
+						map[string]any{
+							"otherField": "bob",
+							"same":       "not same value",
+						},
+					},
+				},
+			},
+			want: want{
+				ignored: []string{"array[0].singleField"},
+			},
+		},
+		"NestedArraySameField": {
+			reason: "Should not the ignored field when init and for provider have nested arrays with same fields",
+			args: args{
+				initProvider: map[string]any{
+					"array": []any{
+						map[string]any{
+							"same": "same",
+						},
+					},
+				},
+				forProvider: map[string]any{
+					"array": []any{
+						map[string]any{
+							"same": "not same value",
+						},
+					},
+				},
+			},
+			want: want{
+				ignored: []string{},
+			},
+		},
```

Comment:
I just tried the following:

<img width="568" alt="Screenshot 2023-07-24 at 14 28 21" src="https://github.com/upbound/upjet/assets/9900707/ae0c9b7c-7401-4bf7-a1d0-789cfec4f621">

But got:

```
    ignored_test.go:270: GetIgnorableFields() got = [array[0].singleField], want [array[0].singleField array[1].newField]
--- FAIL: TestGetIgnoredFields (0.00s)
```

Reply:
good catch, fixed. The result should be 
```go 
[]string{"array[0].singleField", "array[1]"}
```

because there is not equivalent `forProvider` array entry, so the whole array element should be ignored.